src/os_posix/os_fs.c

```diff
@@ -420,9 +431,60 @@ __posix_file_read(
               "%s: handle-read: pread: failed to read %" WT_SIZET_FMT " bytes at offset %" PRIuMAX,
               file_handle->name, chunk, (uintmax_t)offset);
     }
+    WT_STAT_CONN_INCRV(session, block_byte_read_syscall, len);
     return (0);
 }
 
+/*
+ * __posix_file_read_mmap --
+ *     Get the buffer from the mapped region.
+ */
+static int
+__posix_file_read_mmap(
+  WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, wt_off_t offset, size_t len, void *buf)
+{
+    WT_FILE_HANDLE_POSIX *pfh;
+    WT_SESSION_IMPL *session;
+
+    session = (WT_SESSION_IMPL *)wt_session;
+    pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
+
+    __wt_verbose(session, WT_VERB_READ,
+      "read-mmap: %s, fd=%d, offset=%" PRId64
+      ","
+      "len=%" PRIu64 ", mapped buffer: %p, mapped size = %" PRIu64 "\n",
+      file_handle->name, pfh->fd, offset, (uint64_t)len, (void *)pfh->mmap_buf,
+      (uint64_t)pfh->mmap_size);
+
+    WT_ASSERT(session, S2C(session)->mmap);
+
+    if (!pfh->mmap_file_mappable)
+        goto syscall;
+
+    /* Indicate that we might be using the mapped area */
+    (void)__wt_atomic_addv32(&pfh->mmap_usecount, 1);
+
+    /*
+     * If the I/O falls outside of the mapped buffer, or the buffer is being resized, we defer to
+     * the regular system call.
+     */
+    if (pfh->mmap_buf != 0 && pfh->mmap_size >= (size_t)offset + len && !pfh->mmap_resizing) {
+
+        memcpy(buf, (void *)(pfh->mmap_buf + offset), len);
+
+        /* Signal that we are done using the mapped buffer. */
+        (void)__wt_atomic_subv32(&pfh->mmap_usecount, 1);
+
+        WT_STAT_CONN_INCRV(session, block_byte_read_mmap, len);
+        return (0);
+    } else {
+        /* Signal that we won't be using the mapped buffer after all. */
+        (void)__wt_atomic_subv32(&pfh->mmap_usecount, 1);
+syscall:
+        return __posix_file_read(file_handle, wt_session, offset, len, buf);
+    }
+}
```

Comment:
You'll get flagged by various static analyzers for having an `else` after a `return`. You might want to use:
```
    if (pfh->mmap_file_mappable) { 
        /* Indicate that we might be using the mapped area */
        (void)__wt_atomic_addv32(&pfh->mmap_usecount, 1);
         
        /*
         * If the I/O falls outside of the mapped buffer, or the buffer is being resized, we defer  
         * to the regular system call.  
         */ 
        if (pfh->mmap_buf != 0 && pfh->mmap_size >= (size_t)offset + len && !pfh->mmap_resizing) {
            
            memcpy(buf, (void *)(pfh->mmap_buf + offset), len);
        
            /* Signal that we are done using the mapped buffer. */
            (void)__wt_atomic_subv32(&pfh->mmap_usecount, 1);
        
            WT_STAT_CONN_INCRV(session, block_byte_read_mmap, len);
            return (0);
        } 

        /* Signal that we won't be using the mapped buffer after all. */
        (void)__wt_atomic_subv32(&pfh->mmap_usecount, 1); 
    }   

    return __posix_file_read(file_handle, wt_session, offset, len, buf);
```